flink-mesos/src/main/java/org/apache/flink/mesos/runtime/clusterframework/MesosResourceManager.java

```diff
@@ -663,7 +666,10 @@ public void taskTerminated(TaskMonitor.TaskTerminated message) {
 			assert(launched != null);
 			LOG.info("Worker {} failed with status: {}, reason: {}, message: {}.",
 				id, status.getState(), status.getReason(), status.getMessage());
-			startNewWorker(launched.profile());
+
+			if (recordFailure()) {
+				startNewWorker(launched.profile());
+			}
```

Comment:
I think we should move the logic whether we can start a new worker or not to the `ResourceManager`. E.g. we could have the following methods in `ResourceManager`
```
protected Collection<ResourceProfile> tryStartNewWorker(ResourceProfile resourceProfile) {
	if (failureRater.exceedMaximumFailureRate()) {
		return Collections.emptyList();
	}

	return startNewWorker(resourceProfile);
}

protected void startNewWorkerIfNeeded(ResourceProfile resourceProfile, int pendingSlots) {
	int currentPendingSlots = pendingSlots;
	while (currentPendingSlots < getNumberRequiredTaskManagerSlots()) {
		final Collection<ResourceProfile> slots = tryStartNewWorker(resourceProfile);

		if (slots.isEmpty()) {
			break;
		}

		currentPendingSlots += slots.size();
	}
}
```

and then we only call `startNewWorkerIfNeeded` from here.